File: SimpleImageViewer/SimpleImageViewerNet/AlbumSideList.cs
using System;
using System.Drawing;
using SpaceVIL.Common;
using SpaceVIL.Decorations;
using SpaceVIL.Core;
using SpaceVIL;
using System.IO;
using System.Collections.Generic;

namespace SimpleImageViewer
{
    public class AlbumSideList : SideArea
    {
        private bool _isInit = false;
        private List<Album> _list = new List<Album>();
        private ListBox _albumList = new ListBox();
        private ButtonCore _addButton = new ButtonCore();

        private WrapGrid _area;
        private PreviewArea _preview;

        public AlbumSideList(CoreWindow handler, Side attachSide, WrapGrid area, PreviewArea preview) : base(handler, attachSide)
        {
            _area = area;
            _preview = preview;
        }

        public override void InitElements()
        {
            base.InitElements();
            if (_isInit)
                return;

            Window.SetBackground(60, 60, 60);

            Label _name = new Label();
            _name.SetFont(DefaultsService.GetDefaultFont(FontStyle.Bold, 28));
            _name.SetText("My Albums");
            _name.SetMargin(0, 0, 0, 0);
            _name.SetMargin(0, 25, 0, 0);
            _name.SetPadding(10, 10, 10, 0);
            _name.SetHeightPolicy(SizePolicy.Fixed);
            _name.SetHeight(70);

            _albumList.SetSelectionVisible(false);
            _albumList.SetVScrollBarVisible(ScrollBarVisibility.AsNeeded);
            _albumList.SetHScrollBarVisible(ScrollBarVisibility.Never);
            _albumList.SetBackground(Color.Transparent);
            _albumList.SetMargin(10, 100, 10, 50);
            _albumList.VScrollBar.SetStyle(Style.GetSimpleVerticalScrollBarStyle());

            Style style = new Style();
            style.Background = Color.Transparent;
            style.Foreground = Color.Black;
            style.Font = DefaultsService.GetDefaultFont();
            style.SetSizePolicy(SizePolicy.Fixed, SizePolicy.Fixed);
            style.SetSize(40, 40);
            style.SetAlignment(ItemAlignment.Right, ItemAlignment.Bottom);
            style.SetTextAlignment(ItemAlignment.Right, ItemAlignment.Bottom);
            style.SetPadding(4, 4, 4, 4);
            style.SetMargin(0, 0, 10, 10);
            style.SetSpacing(0, 0);
            style.SetBorder(new Border(Color.Transparent, new CornerRadius(20), 0));
            style.AddItemState(ItemStateType.Hovered, new ItemState(Color.FromArgb(100, 255, 255, 255)));
            style.AddItemState(ItemStateType.Pressed, new ItemState(Color.FromArgb(100, 0, 0, 0)));
            _addButton.SetStyle(style);

            CustomShape plus = new CustomShape(GraphicsMathService.GetCross(30, 30, 4, 0));
            plus.SetAlignment(ItemAlignment.HCenter, ItemAlignment.VCenter);
            plus.SetBackground(100, 100, 100);
            plus.SetSizePolicy(SizePolicy.Expand, SizePolicy.Expand);

            AddItems(
                _name,
                _albumList,
                _addButton
            );

            _addButton.AddItem(plus);

            _addButton.EventMouseClick += (sender, args) =>
            {
                Hide();
                OpenEntryDialog browse = new OpenEntryDialog("Open Folder:", FileSystemEntryType.Directory, OpenDialogType.Open);
                browse.OnCloseDialog += () =>
                {
                    if (browse.GetResult() == null || browse.GetResult().Equals(String.Empty))
                    {
                        Show();
                        return;
                    }
                    Album album = new Album(new DirectoryInfo(browse.GetResult()).Name, browse.GetResult());
                    AddAlbum(album);
                    album.OnDoubleClick += (s) =>
                    {
                        Hide();
                        Model.FillImageArea(GetHandler(), _area, _preview, (s as Album).GetPath());
                    };
                };
                browse.Show(GetHandler());
            };

            if (_list.Count > 0)
                foreach (Album item in _list)
                {
                    _albumList.AddItem(item);
                    item._expand.EventToggle += (sender, args) =>
                    {
                        _albumList.GetWrapper(item).UpdateSize();
                        _albumList.GetArea().UpdateLayout();
                    };
                }

            _isInit = true;
        }

        public void AddAlbum(Album album)
        {
            if (_isInit)
            {
                _albumList.AddItem(album);
                album._expand.EventToggle += (sender, args) =>
                {
                    _albumList.GetWrapper(album).UpdateSize();
                    _albumList.GetArea().UpdateLayout();
                };
            }
            else
                _list.Add(album);
        }


        public bool IsHide = true;

        public override void Show()
        {
            IsHide = false;
            base.Show();
        }

        public override void Hide()
        {
            IsHide = true;
            base.Hide();
        }
    }
}

File: Complex/Demos/SimpleImageViewer/SimpleImageViewerJVM/src/main/java/SimpleImageViewerJVM/AlbumSideList.java
package SimpleImageViewerJVM;

import java.awt.Color;
import java.awt.Font;
import java.util.LinkedList;
import java.util.List;
import java.io.File;

import com.spvessel.spacevil.*;
import com.spvessel.spacevil.Common.*;
import com.spvessel.spacevil.Decorations.Border;
import com.spvessel.spacevil.Decorations.CornerRadius;
import com.spvessel.spacevil.Decorations.ItemState;
import com.spvessel.spacevil.Decorations.Style;
import com.spvessel.spacevil.Flags.*;
import com.spvessel.spacevil.Flags.OpenDialogType;

public class AlbumSideList extends SideArea {
    private boolean _isInit = false;
    private List<Album> _list = new LinkedList<Album>();
    private ListBox _albumList = new ListBox();
    private ButtonCore _addButton = new ButtonCore();

    private WrapGrid _area;
    private PreviewArea _preview;

    public AlbumSideList(CoreWindow handler, Side attachSide, WrapGrid area, PreviewArea preview) {
        super(handler, attachSide);
        _area = area;
        _preview = preview;
    }

    @Override
    public void initElements() {
        super.initElements();
        if (_isInit)
            return;

        window.setBackground(60, 60, 60);

        Label _name = new Label();
        _name.setFont(DefaultsService.getDefaultFont(Font.BOLD, 28));
        _name.setText("My Albums");
        _name.setMargin(0, 0, 0, 0);
        _name.setMargin(0, 25, 0, 0);
        _name.setPadding(10, 10, 10, 0);
        _name.setHeightPolicy(SizePolicy.FIXED);
        _name.setHeight(70);

        _albumList.setSelectionVisible(false);
        _albumList.setVScrollBarPolicy(VisibilityPolicy.AS_NEEDED);
        _albumList.setHScrollBarPolicy(VisibilityPolicy.NEVER);
        _albumList.setBackground(new Color(0, 0, 0, 0));
        _albumList.setMargin(10, 100, 10, 50);
        _albumList.vScrollBar.setStyle(Style.getSimpleVerticalScrollBarStyle());

        Style style = new Style();
        style.background = new Color(0, 0, 0, 0);
        style.foreground = Color.BLACK;
        style.font = DefaultsService.getDefaultFont();
        style.setSizePolicy(SizePolicy.FIXED, SizePolicy.FIXED);
        style.setSize(40, 40);
        style.setAlignment(ItemAlignment.RIGHT, ItemAlignment.BOTTOM);
        style.setTextAlignment(ItemAlignment.RIGHT, ItemAlignment.BOTTOM);
        style.setPadding(4, 4, 4, 4);
        style.setMargin(0, 0, 10, 10);
        style.setSpacing(0, 0);
        style.setBorder(new Border(new Color(0, 0, 0, 0), new CornerRadius(20), 0));
        style.addItemState(ItemStateType.HOVERED, new ItemState(new Color(255, 255, 255, 100)));
        style.addItemState(ItemStateType.PRESSED, new ItemState(new Color(0, 0, 0, 100)));
        _addButton.setStyle(style);

        CustomShape plus = new CustomShape(GraphicsMathService.getCross(30, 30, 4, 0));
        plus.setAlignment(ItemAlignment.HCENTER, ItemAlignment.VCENTER);
        plus.setBackground(100, 100, 100);
        plus.setSizePolicy(SizePolicy.EXPAND, SizePolicy.EXPAND);

        addItems(_name, _albumList, _addButton);

        _addButton.addItem(plus);

        _addButton.eventMouseClick.add((sender, args) -> {
            hide();
            OpenEntryDialog browse = new OpenEntryDialog("Open Folder:", FileSystemEntryType.DIRECTORY,
                    OpenDialogType.OPEN);
            browse.onCloseDialog.add(() -> {
                if (browse.getResult() == null)
                {
                    show();
                    return;
                }
                Album album = new Album(new File(browse.getResult()).getName(), browse.getResult());
                addAlbum(album);
                album.onDoubleClick.add((s) -> {
                    hide();
                    Model.fillImageArea(getHandler(), _area, _preview, ((Album) s).getPath());
                });
            });
            browse.show(getHandler());
        });

        if (_list.size() > 0)
            for (Album item : _list) {
                _albumList.addItem(item);
                item._expand.eventToggle.add((sender, args) -> {
                    _albumList.getWrapper(item).updateSize();
                    _albumList.getArea().updateLayout();
                });
            }

        _isInit = true;
    }

    public void addAlbum(Album album) {
        if (_isInit) {
            _albumList.addItem(album);
            album._expand.eventToggle.add((sender, args) -> {
                _albumList.getWrapper(album).updateSize();
                _albumList.getArea().updateLayout();
            });
        } else
            _list.add(album);
    }

    public boolean isHide = true;

    public void show() {
        super.show();
        isHide = false;
    }

    public void hide() {
        super.hide();
        isHide = true;
    }
}